newton/utils/contact_sensor.py

```diff
@@ -0,0 +1,397 @@
+
+        # add contribution for shape a and b
+        for i in range(2):
+            mono_sp = wp.vec2i(-1, pair[i])
+            mono_ord = bisect_shape_pairs(sp_sorted, num_sp, mono_sp)
+
+            # for shape vs all, only one accumulator is supported and flip is trivially true
+            if mono_ord < num_sp and sp_sorted[mono_ord] == mono_sp:
+                force_acc = sp_ep[sp_ep_offset[mono_ord]][0]
+                wp.atomic_add(net_force, force_acc, wp.where(bool(i), -force, force))
+
+
+def _lol_to_arrays(list_of_lists: list[list], dtype) -> tuple[wp.array, wp.array, wp.array]:
+    """Convert a list of lists to three warp arrays containing the values, offsets and counts.
+    Does nothing and returns None, None, None if the list is empty.
+    """
+    if not list_of_lists:
+        return None, None, None
+    a = wp.array([el for l in list_of_lists for el in l], dtype=dtype)
+    count_list = list(map(len, list_of_lists))
+    offset = wp.array(np.cumsum([0, *count_list[:-1]]), dtype=wp.int32)
+    count = wp.array(count_list, dtype=wp.int32)
+    return a, offset, count
+
+
+def convert_contact_info(
+    model: Model,
+    contact_info: ContactInfo,
+    solver: SolverBase | None = None,
+    contacts: Contacts | None = None,
+):
+    """Populate ContactInfo object from the solver or from the Contacts object."""
+    if solver is not None:
+        if isinstance(solver, MuJoCoSolver):
+            solver.update_newton_contacts(model, solver.mjw_data, contact_info)
+        else:
+            raise NotImplementedError("Contact conversion not yet implemented for this solver")
+    if contacts is not None:
+        raise NotImplementedError("Contact conversion not yet implemented for Contacts object")
+
+
+class ContactView:
+    """A view for querying contacts between entities in the simulation.
+    This class stores the parameters of the query and provides a view of the results.
+    """
+
+    def __init__(self, query_id: int, args: dict[str, Any]):
+        self.query_id: int = query_id
+        self.args: dict[str, Any] = args
+        self.finalized: bool = False
+        self.shape: tuple[int, int] = None
+
+        self.net_force: wp.array(dtype=wp.vec3) = None  # force matrix, aliased to contact reporter
+        """Net force matrix, shape (n_sensors, n_contact_partners [+1 if total included])"""
+
+        self.sensor_keys: list[str] = None
+        """Keys for the sensors in the query, n_sensors"""
+        self.contact_partner_keys: list[str] = None
+        """Keys for the contact partners in the query, n_contact_partners"""
+        self.sensor_entities: list[Entity] = None
+        """Entities for the sensors in the query, n_sensors"""
+        self.contact_partner_entities: list[Entity] = None
+        """Entities for the contact partners in the query, n_contact_partners"""
+        self.entity_pairs: np.ndarray = None  # entity pair matrix
```

Comment:
We have a typedef for Numpy arrays in `newton.core.types` called `nparray` because the correct type alias for Numpy arrays is more complicated than this.